Clean tests up

ocefpaf opened the pull request:
This PR aims to:

- make the functions and variables names consistent and meaningful. (For example: `test_build_face_face_connectivity2` &rarr; `test_build_face_face_connectivity_big`.);
- writes and reads all data from a canonical directory no matter where the test is being called;
- use `pytest.fixture` for the grid creations;
- aggregate the fixtures functions in `utilities.py`;
- clean up the temporary files after test;
- fix typos;
- coverage test;
- mane clean-ups;
- remove unnecessary print statements;
- fix #129.

TODO (in a future PR):
- [ ] split all `from_ncfile` and `from_ncdataset` tests (needed by #125);
- [ ] break big tests with several assertions into small unittests;
- [ ] identify and remove redundant test (for example `test_util.py`);
- [ ] #121 
- [ ] investigate if `test_ugrid.py`  is needed. (This test creates a class and test it there.)

PS: I rebased this PR with #120 and #130

ChrisBarker-NOAA commented on test/test_add_data.py:
```diff
@@ -28,14 +28,13 @@ def test_add_all_data():
     depths.attributes['units'] = 'm'
-    depths.attributes["standard_name"] = "sea_floor_depth"
-    depths.attributes["positive"] = "down"
+    depths.attributes['standard_name'] = 'sea_floor_depth'
```
Is there some standard I don't know about for when to use single vs double quotes?

PEP8 says:
"""
In Python, single-quoted strings and double-quoted strings are the same. This PEP does not make a recommendation for this. Pick a rule and stick to it.
"""

so what's the rule for this project?

ocefpaf replied:
> so what's the rule for this project?

I don't have any strong feelings about this.  I only changed because I write a lot of JSON (where I need the double quotes) and python strings (where they don't matter) in a few of my projects.  So I avoid mixing them to make it clear in my head when.  I can revert those if you want.

ChrisBarker-NOAA replied:
nope -- I don't care. I suppose we should be consistent, but I've actually never found it annoying or confusing to be arbitrary about it. I guess because my editor color-codes strings anyway.

ChrisBarker-NOAA commented on test/test_find_uvar.py:
```diff
@@ -114,15 +120,16 @@ def test_two_triangles_without_faces():
+    with chdir(test_files):
+        grid.save_as_netcdf(fname)
+        ug = UGrid.from_ncfile(fname, load_data=True)
+        os.remove(fname)
```
maybe we should write all the generated files to one location, then have a break down fixture that cleans it all up.

One of the tricks with this is that if a test is failing, it's nice to have the generated files around to look at.

I haven't settled on a canonical way to handle this yet myself.

ocefpaf replied:
Let me think about this in another PR.  In fact, if such feature is desired, we can add as a py.test option.  Because we will want to do that locally and not in the CI.

ChrisBarker-NOAA replied:
Why not? in the CI it really doesn't matter if you clean up after yourself or not, as teh environment will get thrown away anyway. (same for conda build, etc...)

When developing, I find it handy for generated files to stick around so I can examine them. And it isn't really a problem to have them stick around, unless someone does the unforgivable `git add *`. Though you do get bunch of files showing up  as untracked by git, which can be annoying.

But this method makes it easy to comment out the `os.remove` when debugging, so not a bad way to go, either. So let's stick with this for now.

ChrisBarker-NOAA commented on test/test_grid_manipulation.py:
```diff
@@ -71,62 +56,41 @@ def test_build_edge_coordinates():
-                                [0.6, 1.1],
-                                [1.6, 1.1]])
+    assert np.allclose(coords, [[1.1, 0.1], [2.6, 1.1],
+                                [2.1, 2.1], [0.6, 1.1], [1.6, 1.1]])
```
yeach! I much prefer the original layout here a logially NX 2 array should look like that in teh file -- much easeir to see what values are what.

ocefpaf replied:
Agreed.  Reverting that.

ChrisBarker-NOAA commented on test/test_grid_manipulation.py:
```diff
@@ -71,62 +56,41 @@ def test_build_edge_coordinates():
-                       (0, 2),
-                       (2, 3),
-                       (1, 3)]
+    grid.boundaries = [(0, 1), (0, 2), (2, 3), (1, 3)]
```
I guess if it can all fit on one like this is better -- maybe. But often you need to add and remove individual points -- it may get to long for a line, and you can add and remove temporarily with a comment if each logical section has it's own line.

ChrisBarker-NOAA commented on test/test_grid_manipulation.py:
```diff
@@ -71,62 +56,41 @@ def test_build_edge_coordinates():
-                                [2.1, 2.1],
-                                [2.6, 1.1]])
+    assert np.allclose(coords, [[1.1, 0.1], [0.6, 1.1],
+                                [2.1, 2.1], [2.6, 1.1]])
```
see above.

ChrisBarker-NOAA commented on test/test_save_as_netcdf.py:
```diff
@@ -78,91 +79,96 @@ def nc_var_has_attr_vals(ds, var_name, att_dict):
+        ds = netCDF4.Dataset(fname)
+        os.remove(fname)
+
+    assert nc_has_variable(ds, 'mesh_2')
```
I'm a bit confused as to how this works -- you open a DataSet against the file you just created, then delete that file. Is the dataset still good? I think it can be on *nix, but I'd expect Windows to get really confused, what with its aggressive file locking and all. but I guess if it does indeed work on Windows then we're OK.

ocefpaf replied:
The whole idea is to avoid conflicting with opened files.

ChrisBarker-NOAA replied:
YUp -- and a good idea that -- I was working on another project can kept getting test failures that I think were different tests trying to read the same netcdf files -- really painful, and I never resolved it. So if this is working, great! Has it been tested on Windows, yet, though?

ChrisBarker-NOAA commented on test/test_save_as_netcdf.py:
```diff
@@ -78,91 +79,96 @@ def nc_var_has_attr_vals(ds, var_name, att_dict):
-
-        assert not nc_var_has_attr(ds, 'mesh_2', "face_edge_connectivity")
+    fname = 'temp.nc'
+    with chdir(test_files):
```
This is a repeated enough pattern it may be worth a utility function.

ocefpaf replied:
I will change this with the fixtures from #132, so let's keep this as is for now if you don't mind.

ChrisBarker-NOAA replied:
sounds good.

ChrisBarker-NOAA commented on test/test_save_as_netcdf.py:
```diff
@@ -182,22 +188,27 @@ def test_write_with_velocities():
+        'location': 'face',
+        'mesh': 'mesh2',
+        })
+    ds.close()
```
if there is a failure, the ds.close() won't get called. In theory, the file should get closed once the name goes out of scope, but I"ve had problems with that -- maybe pytest keeps references around?

Anyway, we should probably cause some failures on purpose to see if this pattern works.

ocefpaf replied:
If I understand the fixture right that is not necessary.  Pytest will take care of that.

ChrisBarker-NOAA commented:
Aside from the style question about multi-line arrays, this all looks good.

(we need a style guide)

But there are merge conflicts, so I can't merge it just yet -- trying to finalize other PRs first.

ocefpaf commented:
> Aside from the style question about multi-line arrays, this all looks good.

Fixed that.

> (we need a style guide)

:+1: 

> But there are merge conflicts, so I can't merge it just yet -- trying to finalize other PRs first.

Rebased and ready for another round.

ChrisBarker-NOAA commented:
OK -- let's merge this baby and see how it shakes out.

I do want to make sure tests all work on Windows -- opening and closing a bunch of netcdf files in tests can be a nightmare.
